// @ngx-formly/core/esm2020/lib/core.mjs
export { FormlyForm } from './components/formly.form';
export { FormlyField } from './components/formly.field';
export { FormlyAttributes as ɵFormlyAttributes } from './templates/formly.attributes';
export { FormlyGroup as ɵFormlyGroup } from './templates/formly.group';
export { FormlyTemplate as ɵFormlyTemplate } from './components/formly.template';
export { FormlyValidationMessage as ɵFormlyValidationMessage } from './templates/formly.validation-message';
export { FORMLY_CONFIG, FormlyConfig } from './services/formly.config';
export { FormlyFormBuilder } from './services/formly.builder';
export { FieldType } from './templates/field.type';
export { FieldArrayType } from './templates/field-array.type';
export { FieldWrapper } from './templates/field.wrapper';
export { FormlyModule } from './core.module';
export { defineHiddenProp as ɵdefineHiddenProp } from './utils';
export { reverseDeepMerge as ɵreverseDeepMerge } from './utils';
export { getFieldValue as ɵgetFieldValue } from './utils';
export { clone as ɵclone } from './utils';
export { observe as ɵobserve } from './utils';
export { isNil as ɵisNil } from './utils';
