// eslint-plugin-ngrx/dist/rules/store/avoid-mapping-selectors.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.messageId = void 0;
const path_1 = __importDefault(require("path"));
const rule_creator_1 = require("../../rule-creator");
const utils_1 = require("../../utils");
exports.messageId = 'avoidMapppingSelectors';
exports.default = (0, rule_creator_1.createRule)({
    name: path_1.default.parse(__filename).name,
    meta: {
        type: 'suggestion',
        ngrxModule: 'store',
        docs: {
            description: 'Avoid mapping logic outside the selector level.',
            recommended: 'warn',
        },
        schema: [],
        messages: {
            [exports.messageId]: 'Map logic at the selector level instead.',
        },
    },
    defaultOptions: [],
    create: (context) => {
        const { identifiers = [] } = (0, utils_1.getNgRxStores)(context);
        const storeNames = identifiers.length > 0 ? (0, utils_1.asPattern)(identifiers) : null;
        if (!storeNames) {
            return {};
        }
        const pipeWithSelectAndMapSelector = `${(0, utils_1.pipeExpression)(storeNames)}:has(CallExpression[callee.name='select'] ~ CallExpression[callee.name='map'])`;
        const selectSelector = `${(0, utils_1.namedCallableExpression)(storeNames)}[callee.object.callee.property.name='select']`;
        return {
            [`:matches(${selectSelector}, ${pipeWithSelectAndMapSelector}) > CallExpression[callee.name='map']`](node) {
                context.report({
                    node,
                    messageId: exports.messageId,
                });
            },
        };
    },
});
